### DISCUSSION TITLE
LSC Smoke Detector PD-8912 Tasmota compatible

### DISCUSSION THREAD

**Discussion by bjoerns1983** [arendst/Tasmota#17767] [General]
I have recently purchased a Wifi Smoke detector form the Brand LSC (Action retailer) for 14€, which is really cheap for a smoke detector with Wifi or Zigbee.
Opened it up and in there is a Tuya TYWE3S Wifi including several pins routed out on the PCB to solder on. So getting Tasmota in that device should be doable.
Another nice guy documented the device really nice on its Website:
https://hessburg.de/wlan-rauchmelder-fuer-1395-eur/

What so i have to do to get the device configured in Tasmota?

**Comment by barbudor** [comment]
As the device is not know in the [template repository](https://templates.blakadder.com/) you are likely the 1st
So it's to you to find out how it is connected, is the sensors connected directly to the ESP or only to the MCU
If you have 2, it could be good to Tasmotize only 1 and keep the other as a reference if needed

Don't forget that with battery operated device it is likely that the ESP will be OFF most of the time to same battery and woke up only when the TuyaMCU needs an Internet connection

**Reply by bjoerns1983** [reply]
Yes the Tuya Module will sleep most of the time. In the manual it also says thta you need o push the button connect the Board to your phone. So i think the module will be powered up from time to time or state changes.

**Reply by bjoerns1983** [reply]
Okay so the Method would be as followed:
1. Back up the Tuya FW (is there any guide for it)
2. Flash Tasmota und configure to use Tasmota MCU
3. Look into the console if something happens if i push the test button


I would say this detector will only send if a alarm is triggerd and battary percentage or low battaery what i could simulate using an adjustable psu. 

Do i need to enable any debug features on the console? Can i forward the console output someware because the module will go to sleep inbetween?

**Reply by barbudor** [reply]
Weblog 3 would print details of the tuya dialog
For this part you should try to prevent the MCU to reset/power down the esp as it will be difficult to analyse what is happening with an ESP always off

**Reply by bjoerns1983** [reply]
So maxbe powering the Tuya Module directly hoping that only power will be cut

**Reply by bjoerns1983** [reply]
Okay using the VCC pads on the pcb is a bad idea, seems the Tuya module gets powered down after maybe 30sec.
I now directly solders 3.3V to the Module and made a bakup using::
esptool.py --bau 115200 --port /dev/ttyUSB0 read_flash 0x000000 0x200000 flash_2M.bin

I read that the Module has 2M of flash is that correct?

**Reply by barbudor** [reply]
> I read that the Module has 2M of flash is that correct?

yes

**Reply by bjoerns1983** [reply]
Okay uploaded the battery firmware through esptool but after the powercyle the module does not open an AP. Am i missing something?

**Reply by barbudor** [reply]
Which exact command line did you used ?

https://tasmota.github.io/docs/Getting-Started/#upload-firmware

Note: the Web installer is so easy and can install the tasmota-batetry firmware automatically
[https://](https://tasmota.github.io/install/)

**Reply by bjoerns1983** [reply]
esptool.py --bau 115200 --port /dev/ttyUSB0 write_flash -fs 1MB -fm dout 0x0 tasmota-battery.bin    

The webinstaller might be great but i am on a Linux Laptop with Firefox where the webinstaller does not work

**Reply by barbudor** [reply]
strange, your line looks ok, although I never use the baud rate parameter

How do you power the device currently ? Some device won't boot correctly and show up the AP if powered from the USB adapter
You may use a serial terminal to have a look at the logs

A pity that firefox doesn't support Web Serial

**Reply by bjoerns1983** [reply]
Okay i just downloaded chromium flashed it (using a CP2102 i now not the best USB Serial Adapter) via the webinstaller and now that i power it via the adjustable PSU  the Wifi shows up.

**Reply by Jason2866** [reply]
Chromium should work. The correct command for speed is `-baud` but i think that's just a typo here. Esptool.py would have not worked...
You need to power the esp8266 with a external 3.3V source to get an AP, since the device circuit will not power the esp8266. There is no smoke alert.

**Reply by bjoerns1983** [reply]
Okay got it in my Wifi and the MCU is also doing something:
![image](https://user-images.githubusercontent.com/424566/213936186-aa8c188f-ca6c-449a-9084-82127c31123d.png)

Nevertheless, i have to bring the children to bed so no more tinkering toad.
Any useful hints how to investigate further tomorrow?

**Reply by bjoerns1983** [reply]
Okay back at the Keyboard. The problems yesterday were because of bad soldering. After i soldered the 3.3V correctly i was able to flash using ESP Tool, just wanted to confirm that.

Is there any good and easy explanation how i can now get the info out of the smoke detector an translate it for tasmota?

**Reply by barbudor** [reply]
Tuya provides sensors data using "DpId" 
Tasmota knows how to map some of them but may be not those used by your device (could be the 1st smoke sensor)
I've heard of a web based tool that can help to understand Tuya message
but you can also analyse them by hand
If needed, we can add more mapping in Tasmota

**Reply by bjoerns1983** [reply]
Problem is that i cant figure out any DpID, also using the Tuya Tasmota Helper.
It looks like the connection between the TYWE3s and the TuyaMCU is there and can communitcate.
Ate least i get a Tuya Recieve and a Product Number after startup:
```
13:50:28.138 {"TuyaReceived":{"Data":"55AA000100247B2270223A22636A6C75746B757576786E696531376F222C2276223A22312E302E30227D17","Cmnd":1,"CmndData":"7B2270223A22636A6C75746B757576786E696531376F222C2276223A22312E302E30227D"}}
13:50:28.139 TYA: MCU Product ID: {"p":"cjlutkuuvxnie17o","v":"1.0.0"}
```
But after that there is not much userful for me:

```
13:57:58.819 CMD: Backlog Weblog 4; TuyaSend8;
13:57:58.821 SRC: WebConsole from 192.168.2.179
13:57:58.823 CMD: Grp 0, Cmd 'BACKLOG', Idx 1, Len 20, Pld -99, Data 'Weblog 4; TuyaSend8;'
13:57:58.843 SRC: Backlog
13:57:58.846 CMD: Grp 0, Cmd 'WEBLOG', Idx 1, Len 1, Pld 4, Data '4'
13:57:58.849 RSL: RESULT = {"WebLog":4}
13:57:59.064 SRC: Backlog
13:57:59.066 CMD: Grp 0, Cmd 'TUYASEND', Idx 8, Len 0, Pld -99, Data ''
13:57:59.069 TYA: Read MCU state
13:57:59.078 TYA: Send "55aa0001000000"
13:57:59.081 RSL: RESULT = {"TuyaSend":"Done"}
13:57:59.219 CMD: TuyaMCU;
13:57:59.221 SRC: WebConsole from 192.168.2.179
13:57:59.224 CMD: Grp 0, Cmd 'TUYAMCU', Idx 1, Len 1, Pld -99, Data ';'
13:57:59.227 TYA: TuyaMcu Invalid function id=0
13:57:59.230 RSL: RESULT = {"TuyaMCU":[{"fnId":11,"dpId":1}]}
13:57:59.422 TYA: Send "55aa00000000ff"
13:58:06.801 WIF: Checking connection...
13:58:10.417 TYA: Send "55aa00000000ff"
13:58:20.926 {"TuyaReceived":{"Data":"55AA0007000006","Cmnd":7}}
13:58:21.402 TYA: Send "55aa00000000ff"
13:58:21.441 {"TuyaReceived":{"Data":"55AA000000010000","Cmnd":0,"CmndData":"00"}}
13:58:21.444 TYA: Heartbeat
13:58:21.445 TYA: Detected MCU restart
13:58:21.915 {"TuyaReceived":{"Data":"55AA0007000006","Cmnd":7}}
13:58:22.420 TYA: Set WiFi LED 3 (0)
13:58:22.430 TYA: Send "55aa000300010306"
13:58:22.929 {"TuyaReceived":{"Data":"55AA0007000006","Cmnd":7}}
13:58:23.449 TYA: Send "55aa001c0008011701170d3a1701b2"
13:58:23.944 {"TuyaReceived":{"Data":"55AA0007000006","Cmnd":7}}
13:58:24.957 {"TuyaReceived":{"Data":"55AA0007000006","Cmnd":7}}
13:58:25.408 WIF: Sending Gratuitous ARP
13:58:25.973 {"TuyaReceived":{"Data":"55AA0007000006","Cmnd":7}}
13:58:26.776 WIF: Checking connection...
13:58:26.987 {"TuyaReceived":{"Data":"55AA0007000006","Cmnd":7}}
13:58:28.001 {"TuyaReceived":{"Data":"55AA0007000006","Cmnd":7}}
13:58:29.016 {"TuyaReceived":{"Data":"55AA0007000006","Cmnd":7}}
13:58:30.031 {"TuyaReceived":{"Data":"55AA0007000006","Cmnd":7}}
13:58:31.050 {"TuyaReceived":{"Data":"55AA0007000006","Cmnd":7}}
13:58:32.064 {"TuyaReceived":{"Data":"55AA0007000006","Cmnd":7}}
13:58:32.423 TYA: Send "55aa00000000ff"
13:58:32.462 {"TuyaReceived":{"Data":"55AA000000010101","Cmnd":0,"CmndData":"01"}}
13:58:32.464 TYA: Heartbeat
13:58:33.079 {"TuyaReceived":{"Data":"55AA0007000006","Cmnd":7}}
13:58:34.093 {"TuyaReceived":{"Data":"55AA0007000006","Cmnd":7}}
13:58:35.108 {"TuyaReceived":{"Data":"55AA0007000006","Cmnd":7}}
13:58:36.123 {"TuyaReceived":{"Data":"55AA0007000006","Cmnd":7}}
13:58:37.138 {"TuyaReceived":{"Data":"55AA0007000006","Cmnd":7}}
13:58:38.152 {"TuyaReceived":{"Data":"55AA0007000006","Cmnd":7}}
13:58:39.166 {"TuyaReceived":{"Data":"55AA0007000006","Cmnd":7}}
13:58:40.182 {"TuyaReceived":{"Data":"55AA0007000006","Cmnd":7}}
13:58:41.196 {"TuyaReceived":{"Data":"55AA0007000006","Cmnd":7}}
13:58:42.211 {"TuyaReceived":{"Data":"55AA0007000006","Cmnd":7}}
13:58:43.226 {"TuyaReceived":{"Data":"55AA0007000006","Cmnd":7}}
13:58:43.434 TYA: Send "55aa00000000ff"
13:58:43.474 {"TuyaReceived":{"Data":"55AA000000010101","Cmnd":0,"CmndData":"01"}}
13:58:43.475 TYA: Heartbeat
```
Normally i thin after the TuyaReceived there should be something like
TYA: fnId=11 is set for dpId=1

The TuyaReceived command ending with 6 comes when i press the Testbutton.

But thats simply not happening.

**Reply by barbudor** [reply]
Another user as reported that his device do not send any DP until the ESP reports "Connected to the cloud" which actually occurs only when Tasmota is connected to a MQTT server
If you intend to and have a MQTT broker ready, could you connect Tasmota to it to see if it makes change ?

Also maybe the device has no DP to transfer until it detects some smoke

**Reply by bjoerns1983** [reply]
Enabling MQTT makes no difference.
Using the Battery build of Tasmota is also a problem. After a certain time the smoke detector complains that there is no Wifi connection by beeping, which is kind of annoying.

With the normal version of Tasmota i have seen that after issuing a test alarm the TuyaMCU does not send the Wifi chip back to sleep which means it would drain the battery.

Also i have no idea how to safely test for a real alert.

The Low Battery alert does not wake the Wifi Chip, also i have seen nothing send via MQTT which could be a Battery level.
Dont know what to do with this piece of hardware.

**Reply by barbudor** [reply]
I can't see any reason why the battery binary would makes a difference. It also has the TuyaMCU driver. It just set some SetOption that won't be enforced unless the binary if flashed on a clean build.

> With the normal version of Tasmota i have seen that after issuing a test alarm the TuyaMCU does not send the Wifi chip back to sleep which means it would drain the battery.

Does the MCU keeps on talking to the ESP after ?

I can't seem to find a Tuya command that would explicitly tell the ESP to go back to sleep by itself but unless the MCU has a way to cut the power to the ESP, it would makes sense that TuyaMCU send a go-to-sleep command.
There is only a reference to "sleep" in the description of command 0x25 (Disable Heartbeats) which is currently not implemented on Tasmota. It just says that the MCU would disable the heartbeats before the ESP module goes to sleep.
But I can't find an actual command to send the module to sleep.

![image](https://user-images.githubusercontent.com/2996042/214083682-389cdd18-77a6-4e93-baa5-045da198f0e8.png)


So have you seen any message about unsupported command such as 0x25 (37) ?

> Also i have no idea how to safely test for a real alert.

If you smoke, you could test with cigarette smoke. Otherwise, you may burn a piece of paper below the device.
Steam may work too (time for a 🍵)


Battery operated Tuya devices are generally the worst of the Tuya family to be tasmotized.
One possible option, if you are in tinkering, is to get rid of the MCU and see if you can hook the ESP directly to the sensor, but that would make it hard to run out of battery.

**Reply by bjoerns1983** [reply]
Okay i think i will give up on this and reflash the backed up Tuya FW.
I managed to get a real alarm but the Wifi Module is not activated after this.

Maybe i give tuyapi a go.

**Comment by Jason2866** [comment]
Read the linked Website. In the Photo a Tuya esp8266 module can be seen. The device goes online only if there is a smoke alert. Save the orig. firmware, maybe needed for further testing. Flash the unofficial Tasmota Battery version.

**Reply by bjoerns1983** [reply]
Maybe a dumb question but what is the special about the Tasmota Battery version?

**Reply by barbudor** [reply]
It's mainly a few options to make Wifi connection faster such as disabling re-scan at start and further re-scan (SO56/SO57 off)
As well as enabling DeepSleep

See details here:
https://github.com/Jason2866/Tasmota-build/blob/d6971835c9559ced7ecde06550f1c66b24e0328c/tasmota/user_config_override.h#L1290-L1331

**Comment by kroon040** [comment]
Maybe temporary  connect the tuya also to the battery. It worked for configuring better (I have a nedis smoke, and that was also needed)
